### PR TITLE
feat(python): Excel export support via new `write_excel` IO method

### LINKED ISSUE #5568: Add DataFrame.write_excel
### Problem description

The read_excel file was added in https://github.com/pola-rs/polars/pull/3567 and it would be nice if there was also a `DataFrame.write_excel` function, maybe using [xlsxwriter-rs](https://github.com/informationsea/xlsxwriter-rs).

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #5568.

Extremely capable Excel export, leveraging `xlsxwriter`; this first iteration supports the following capabilities:

* Export to the filesystem, BytesIO, or write into an existing/open `xlsxwriter.Workbook`.
* Write table to any position on the worksheet, using excel notation like `"B3"` or zero-indexed`(row,col)` coordinates.
* Support named table styles, with optional style modifiers.
* Name the exported table object so it can be targeted by subsequent sheet operations.
* Per-dtype formatting (also supporting dtype _groups_ like `INTEGER_DTYPES`).
* Per-column formatting (can override dtype format and/or change colours, font, etc).
* Conditional formatting; simple/default format types, as well as icon sets, formulae, etc.
* Control over the default floating point precision (does not round the underlying data).
* Autofit columns (can be overridden per-column by explicit setting `column_widths`).
* Optionally write table without any column headers.
* Includes table autofilter by default (unless `has_headers` is False).
* Optionally hide named columns.
* Optionally hide worksheet grid lines.

## Notes

* If writing into an existing/open `xlsxwriter.Workbook` can further integrate arbitrary objects/formatting with the exported tables (eg: add graphs that reference the named table, add titles/text, write multiple frames/slices to one or more worksheets, etc).

* The `write_excel` method handles all table-relative range calculation so that you can define all formatting by column name/type; you do not need to track sheet position/range (presence/absence of column headers/totals and any initial position offset is all accounted for).

## Examples

* **Test frame**
  ```python
  from datetime import date
  import polars as pl
  
  df = pl.DataFrame(
      {
          "dtm": [date(2023, 1, 1), date(2023, 1, 2), date(2023, 1, 3)],
          "str": ["aaa", "bbb", "ccc"],
          "val": [100.5, 55.0, -99.5],
      }
  )
  ```
* **Basic/default export**
  _Default output includes autofilter and default datetime/date, int, and float formats._
  ```python
  df.write_excel( autofit=True )
  ```
  xl_export_basic
  
* **Simple custom formatting**
  _Apply a named table style, a total row, override default float precision._
  ```python
  df.write_excel(
      table_style = "Table Style Medium 4", 
      float_precision = 6, 
      column_totals = True, 
      autofit = True,
  )
  ```
  xl_export_basic
  
* **Advanced/conditional formatting** 
  _Apply more sophisticated/customised formatting, integrating frame export with xlsxwriter to write multiple tables/frames to the sheet and add additional elements (custom-formatted table titles in this case, but could also be charts/etc)._
  ```python
  from xlsxwriter import Workbook
  
  with Workbook("multi_frame.xlsx") as wb:
      # -------------------------------------
      # basic/default conditional formatting
      # -------------------------------------
      df.write_excel(
        workbook = wb,
        worksheet = "data",
        position = (3, 1),  # specify position as (row,col) coordinates
        table_style = "Table Style Light 20",
        conditional_formats = {"val": "data_bar"},
      )
      
      # -----------------------------------------------
      # advanced conditional formatting, custom styles
      # -----------------------------------------------
      df.write_excel(
        workbook = wb,
        worksheet = "data",
        position = (len(df) + 7, 1),  # calculate row-coordinate offset
        table_style = {
            "style": "Table Style Light 20",
            "first_column": True,
        },
        column_formats={
            "dtm": {"font_color": "#31869c", "bg_color": "#b7dee8"},
        },
        conditional_formats={
            "dtm": {
                "type": "icon_set", 
                "icon_style": "3_traffic_lights",
            },
            "val": {
                "type": "data_bar",
                "data_bar_2010": True,
                "bar_color": "#4bacc6",
                "bar_border_color": "#31869c",
                "bar_negative_color": "#f79745",
                "bar_negative_border_color": "#e26b09",
            },
        },
        column_widths = {"val": 125},
        hide_gridlines = True,
        autofit=True,
      )
      
      # --------------------------------------------------------------------
      # use xlsxwriter integration to add table titles (with custom format)
      # --------------------------------------------------------------------
      ws = wb.get_worksheet_by_name("data")
      fmt_title = wb.add_format(
        {
            "font_color": "#31869c",
            "font_size": 12,
            "italic": True,
            "bold": True,
        }
      )
      ws.write(2, 1, "Basic/default conditional formatting", fmt_title)
      ws.write(len(df)+6, 1, "Customised conditional/table formatting", fmt_title)
    
  ```
  xl_export_advanced

### CODE REVIEW

File: py-polars/polars/datatypes.py [hunk 1]
```diff
@@ -71,7 +73,12 @@ def get_args(tp: Any) -> Any:
 # act as specialisations (eg: "List" and "List[Int32]")
 PolarsDataType: TypeAlias = Union["DataTypeClass", "DataType"]
 PolarsTemporalType: TypeAlias = Union[Type["TemporalType"], "TemporalType"]
-
+OneOrMoreDataTypes: TypeAlias = Union[
+    PolarsDataType,
+    Sequence[PolarsDataType],
+    FrozenSet[PolarsDataType],
+    Set[PolarsDataType],
```
Comment (stinodego) [inline py-polars/polars/datatypes.py:80]: Why not use Iterable here?

```suggestion
    Iterable[PolarsDataType],
```

**Comment by stinodego** [comment]
This would be great to have! And very neat implementation as always 👍

File: py-polars/polars/internals/io.py [hunk 2]
```diff
@@ -23,8 +32,22 @@
     from polars.polars import parquet_schema as _parquet_schema
 
 if TYPE_CHECKING:
+    from xlsxwriter import Workbook  # type: ignore[import]
+    from xlsxwriter.worksheet import Worksheet  # type: ignore[import]
```
Comment (stinodego) [inline py-polars/polars/internals/io.py:36]: Why is this import ignored? If `xlsxwriter` has no type hints, we should just add it to the `pyproject.toml` as a module that should be ignored.

Comment (alexander-beedie) [reply py-polars/polars/internals/io.py]: Apparently I was distracted by the `xlsx2csv` import doing the same thing - have moved both into `pyproject.toml` now ;)

File: py-polars/polars/datatypes.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/datatypes.py]: Apparently because I shouldn't type-hint at midnight 🤣 
Fixed!

**Comment by alexander-beedie** [comment]
> This would be great to have! 

Shall we add the `highlight` tag to this one for the changelog?
I can imagine it being quite popular.

**Comment by ritchie46** [comment]
> Shall we add the `highlight` tag to this one for the changelog? I can imagine it being quite popular.

To you the honors. ;)

File: py-polars/polars/internals/io.py [hunk 3]
```diff
@@ -250,3 +273,145 @@ def _update_columns(df: pli.DataFrame, new_columns: list[str]) -> pli.DataFrame:
         new_columns = cols
     df.columns = new_columns
     return df
+
+
+def _xl_setup_workbook(
```
Comment (ritchie46) [inline py-polars/polars/internals/io.py:278]: Can we make a separate `io_excel` module all excel related code? This is becoming more than a wrapper to our rust backends. That will keep the logic separated.

**Review by ritchie46 — COMMENTED** [review]
Wow, and so many options immediately. Great stuff.

File: py-polars/polars/internals/io.py (hunk 3 above)
Comment (alexander-beedie) [reply py-polars/polars/internals/io.py:278]: Makes sense; will take care of it ;)

**Comment by alexander-beedie** [comment]
(Squashed/rebased)

**Comment by ritchie46** [comment]
Thanks @alexander-beedie. This will likely make a lot of people happy.

**Pull request merged**